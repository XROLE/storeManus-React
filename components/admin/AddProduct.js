import React, { Fragment, Component } from 'react';
import { Link } from 'react-router-dom';
import { ToastContainer, toast } from 'react-toastify';
import { connect } from 'react-redux';
import Spinner from 'react-spinkit';
import Navbar from '../navbar/Navbar';
import Sidebar from '../sidebar/Sidebar';
import TopBar from '../sidebar/TopBar';
import 'react-toastify/dist/ReactToastify.css';
import { addProduct } from '../../store/actions/actions';

class AddProduct extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      price: '',
      quantity: '',
      type: '',
      category: '',
      shouldShowError: false,
      showSuccessMessage: false,
      resetState: false,
    };
  }

  onChange(e) {
    return this.setState({
      [e.target.name]: e.target.value,
    });
  }

  handleSubmit(e) {
    e.preventDefault();
    const product = this.state;
    const { add } = this.props;

    this.setState({
      shouldShowError: true,
      showSuccessMessage: true,
    });
    add(addProduct(product));
  }

  notify(error) {
    toast.error(error);
  }

  render() {
    const {
      name, type, price, category, quantity, shouldShowError, showSuccessMessage,
    } = this.state;
    const {
      pending, addProductError, success, addProductMessage,
    } = this.props;

    if (addProductError && shouldShowError) {
      this.notify(addProductError);
      this.setState({
        shouldShowError: false,
      });
    }

    if (success && showSuccessMessage) {
      this.notify(addProductMessage);
      this.setState({
        name: '',
        price: '',
        quantity: '',
        type: '',
        category: '',
        shouldShowError: false,
        showSuccessMessage: false,
      });
    }
    return (
      <Fragment>
        <Navbar />
        <div className="dasboardDiv">
          <Sidebar />
          <div className="dashboardDisplayDiv">
            <TopBar />
            <div className="dasboard-display-div">
              <form action="" id="add-product" onSubmit={e => this.handleSubmit(e)}>
                <div className="form-label">
                  <label htmlFor="add-product" className="fontBlack">
                    <i className="far fa-plus-square" />
                    {' '}
&nbsp;ADD PRODUCT
                  </label>
                </div>
                <div>
                  <input type="text" name="name" id="product-Name" placeholder="NAME" onChange={e => this.onChange(e)} value={name} />
                </div>
                <div>
                  <input type="number" name="quantity" id="product-Quantity" placeholder="QUANTITY" onChange={e => this.onChange(e)} value={quantity} />
                </div>
                <div>
                  <input type="number" name="price" id="product-Price" placeholder="PRICE" onChange={e => this.onChange(e)} value={price} />
                </div>
                <div>
                  <input type="text" name="type" id="product-Type" placeholder="TYPE" onChange={e => this.onChange(e)} value={type} />
                </div>
                <div>
                  <input type="text" name="category" id="product-Category" placeholder="CATEGORY" onChange={e => this.onChange(e)} value={category} />
                </div>
                <div>
                  <button className="link" id="add-product-button" type="submit">Add Product</button>
                </div>
                { pending === true && (
                <Spinner name="circle" className="spinner" id="reactLoader" />
                )}
              </form>
            </div>
          </div>
        </div>
        <ToastContainer />
      </Fragment>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  add: product => dispatch(product),
});

const mapStateToProps = state => ({
  addProductMessage: state.addProduct.message,
  pending: state.addProduct.pending,
  addProductError: state.addProduct.error,
  success: state.addProduct.success,
});

export default connect(mapStateToProps, mapDispatchToProps)(AddProduct);
