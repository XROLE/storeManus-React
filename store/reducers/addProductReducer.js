import {
  ADD_PRODUCT_IN_PROGRESS,
  ADD_PRODUCT_SUCCESS,
  ADD_PRODUCT_ERROR,
} from '../actions/actionTypes';

const initialState = {
  message: '',
  pending: false,
  error: '',
};

export default function (state = initialState, action) {
  switch (action.type) {
    case ADD_PRODUCT_IN_PROGRESS: {
      return {
        ...state,
        error: null,
        pending: true,
      };
    }
    case ADD_PRODUCT_SUCCESS: {
      return {
        ...state,
        pending: false,
        success: action.payload.success,
        message: action.payload.message,
      };
    }
    case ADD_PRODUCT_ERROR: {
      return {
        ...state,
        error: action.error,
        pending: false,
        success: action.success,
      };
    }
    default:
      return state;
  }
}
