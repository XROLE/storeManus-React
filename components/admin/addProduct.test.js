import React from 'react';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import axios from 'axios';
import { BrowserRouter } from 'react-router-dom';
import AddProduct from './AddProduct';
import addProductReducer from '../../store/reducers/addProductReducer';

jest.mock('axios');

describe('Admin Add Products', () => {
  it('Should render just the Admin Add Attendant component', () => {
    const component = shallow(
      <AddProduct />,
    );
    expect(component.exists()).toBe(true);
    expect(component).toMatchSnapshot();
  });
});

describe('Add Product Component', () => {
  const add = jest.fn();
  const initialState = {
    addProduct: {
      message: '',
      pending: false,
      error: '',
    },
  };
  const mockStore = configureStore([thunk]);
  const store = mockStore(() => initialState);
  const dispatchMock = jest.spyOn(store, 'dispatch');
  const addProductMessage = '';
  const addProductError = '';
  const pending = false;
  const success = '';

  const props = {
    registerAttendant: {
      add,
      addProductMessage,
      addProductError,
      pending,
      success,
    },
  };

  const wrapper = mount(
    <Provider store={store}>
      <BrowserRouter>
        <AddProduct {...props} />
      </BrowserRouter>
    </Provider>,
  );

  it('should call the add function', () => {
    wrapper.find('#product-Name').simulate(
      'change', {
        target: {
          name: 'name', value: 'coffee',
        },
      },
    );

    wrapper.find('#product-Quantity').simulate(
      'change', {
        target: {
          name: 'quantity', value: 50,
        },
      },
    );

    wrapper.find('#product-Price').simulate(
      'change', {
        target: {
          name: 'price', value: 30,
        },
      },
    );

    wrapper.find('#product-Type').simulate(
      'change', {
        target: {
          name: 'type', value: 'Tea',
        },
      },
    );

    wrapper.find('#product-Category').simulate(
      'change', {
        target: {
          name: 'category', value: 'Beverage',
        },
      },
    );

    axios.post.mockImplementation(() => Promise.resolve({
      data: { Success: true, Message: 'profile updated successfully' },
    }));

    wrapper.find('form').simulate(
      'submit',
      {
        preventDefault: () => {},
      },
    );
    expect(dispatchMock).toHaveBeenCalled();
  });

  it('should cnot add product when a field is empty', () => {
    wrapper.find('#product-Name').simulate(
      'change', {
        target: {
          name: 'name', value: '',
        },
      },
    );

    wrapper.find('#product-Quantity').simulate(
      'change', {
        target: {
          name: 'quantity', value: 50,
        },
      },
    );

    wrapper.find('#product-Price').simulate(
      'change', {
        target: {
          name: 'price', value: 30,
        },
      },
    );

    wrapper.find('#product-Type').simulate(
      'change', {
        target: {
          name: 'type', value: 'Tea',
        },
      },
    );

    wrapper.find('#product-Category').simulate(
      'change', {
        target: {
          name: 'category', value: 'Beverage',
        },
      },
    );

    const error = {
      response: {
        data: {
          Message: 'No empty field is allowed',
          Success: false,
        },
      },
    };
    axios.post.mockImplementation(() => Promise.reject(error));

    wrapper.find('form').simulate(
      'submit',
      {
        preventDefault: () => {},
      },
    );
    expect(dispatchMock).toHaveBeenCalled();
  });
});

it('should setup default state values', () => {
  const state = addProductReducer(undefined, {
    type: '@@INIT',
  });
  expect(state).toEqual({
    message: '',
    pending: false,
    error: '',
  });
});

it('should update state when action type is  ADD_PRODUCT_IN_PROGRESS', () => {
  const state = addProductReducer(undefined, {
    type: 'ADD_PRODUCT_IN_PROGRESS',
  });
  expect(state).toEqual({
    ...state,
    pending: true,
  });
});

it('should update state when action type is  ADD_PRODUCT_SUCCESS', () => {
  const state = addProductReducer(undefined, {
    type: 'ADD_PRODUCT_SUCCESS',
    payload: {
      success: true,
    },
  });
  expect(state).toEqual({
    ...state,
    success: true,
  });
});

it('should update state when action type is  ADD_PRODUCT_ERROR', () => {
  const state = addProductReducer(undefined, {
    type: 'ADD_PRODUCT_ERROR',
    error: 'failed',
  });
  expect(state).toEqual({
    ...state,
    error: 'failed',
  });
});
