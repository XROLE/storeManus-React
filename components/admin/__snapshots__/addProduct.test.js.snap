// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin Add Products Should render just the Admin Add Attendant component 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
